### ISSUE TITLE
[vcpkg] Add a optional "Removes:" control field for ports

### ISSUE THREAD

**Issue by Neumann-A** [microsoft/vcpkg#8126]
I would like to see an optional `Removes:` control field in the `CONTROL` file of the ports. Reason: metaports like qt5 or openssl. In this control field ports are given which will be removed alongside the requested port and it should follow the same syntax and logic as `Build-Depends:`.  The problem with metaports is that they are able to install all there dependencies but are missing a way to also remove them. So currently `vcpkg remove qt5` just removes the qt5 metaport but not the whole qt5 suite. In the case of qt5 the control field could be `Removes: qt5-base` which then removes all qt5 related ports since they will be all dependent on qt5-base. So `vcpkg remove qt5` would effectively translate to `vcpkg remove qt5 qt5-base` (in this case probably also requires a --recurse option. As an alternative list all ports to be removed in the `Removes:` field)

**Comment by cbezault** [comment]
I like this, we'll discuss internally and decide if this is the path forward.

**Comment by heydojo** [comment]
With all due respect @Neumann-A your strategy appears flawed imo.
One of the primary lessons learned from Linux package management over the last twenty or so years is that dependency hell is a thing and trying to unpick a deeply entwined dependency tree creates deep running complex problems.

I propose some sort of derivation approach to the problem.

Please read Eelco Dolstra's Phd thesis on Nix (The purely functional package manager.) https://nixos.org/~eelco/pubs/phd-thesis.pdf to understand the approach (if you have not already.)

I believe that the approach Eelco has taken to package management is correct and having read the thesis when (over ten years ago) it was first released, time has proven as much.
There are lessons to be learned from Dr Dolstra's thesis which apply to vcpkg. In particular, how dependency hell is avoided.

Consider this:

- Package A depends upon openssl 1.1.0
- Package B depends upon openssl 1.1.1 but will not compile against openssl 1.1.0 because it relies on feature x from openssl 1.1.1 which is unavailable in openssl 1.1.0

If you want to include package A within vcpkg's ports tree with a dependency upon openssl, then package B cannot be added to vcpkg with a dependency upon the same openssl port because two different versions are required by the two separate packages.
To accommodate the issue, openssl may then be forked out into openssl and openssl-lts within the tree.
Thus, there becomes a split and a remove function is not the solution. Overlapping packages which contain the same or similar files but have been forked out become like spaghetti.

In short, imo you need a gc (garbage collector) function or daemon routine to deal with a "remove" clean up instead of a remove declaration and function IMO. And the storage of each resulting port's compile must be unique, at the very least in terms of it's storage identification.

I'd also like to recommend that some dependency graphs are published (hopefully publicly along with the project) so that developers can visualise the stats surrounding how dependencies and their relationships are handled within vcpkg.

Sorry that was a lot. Thanks for reading.

**Comment by Neumann-A** [comment]
There will only by one openssl port installed ever. If it is worked out into different ports `vcpkg remove openssl` should remove all of them. If you want to remove a specific one use `vcpkg remove openssl-whatever`

>Overlapping packages which contain the same or similar files but have been forked out become like spaghetti.

1. there is no spaghetti. even if ports are fork out into different ports the base (meta)port can only be installed **once** in vcpkg. 
2. due to 1. you should not be able to install different versions of the same port twice without getting an error. So in your example vcpkg basically tells you to either: choose another package manager, update Package A to the new openssl version or use another vcpkg instance to build your package A.

>Please read Eelco Dolstra's Phd thesis on Nix (The purely functional package manager.) https://nixos.org/~eelco/pubs/phd-thesis.pdf to understand the approach (if you have not already.)

sry not going to read a >250 pages thesis which is not in my field and which seems to deal with a rather simple man made problem. (I skipped over it and the simplest solution is to keep everything on one version - period - which is basically what vcpkg is trying to achieve.)

But I fail to see why you are against the `Removes:` field. It's intended use is not to remove conflicting ports but to remove metaports which are in general a collection of other ports like boost. You do `vcpkg install boost` and all boost libraries get installed. If you do `vcpkg remove boost` only the boost metaport gets removed and the rest of boost is still installed. The requested feature is just to fix that behavior so that  `vcpkg remove boost` removes every boost port which is in my eyes what I as the user would expect it to do.

**Comment by Neumann-A** [comment]
@cbezault: or maybe better call it `Remove-Depends` without listing any ports and just use whatever `Build-Depends` were used in building the port.
